File: compilar.zig
const std = @import("std");
const c = @cImport({
    @cInclude("cmark.h");
});
const endsWith = std.mem.endsWith;

fn stripExtension(file_name: []const u8) ![]const u8 {
    const index = std.mem.lastIndexOfLinear(u8, file_name, ".");
    if (index) |i| {
        return file_name[0..i];
    } else return error.NoExtension;
}

const HeaderOptions = struct {
    // Si esto es true, se muestra un botón para volver al index.
    ir_al_inicio: bool = true,
    // Si esto es true, se muestra un <header>.
    header: bool = true,
    // Si esto es true, se muestra un link a #conexiones en el header.
    has_connections: bool = false,
};
fn header(
    writer: std.fs.File.Writer,
    title: []const u8,
    src_name: []const u8,
    options: HeaderOptions,
) !void {
    // TODO: deshardcodear base_uri
    try writer.print(
        \\<!doctype html>
        \\<meta charset=utf-8>
        \\<meta name=viewport content="width=device-width, initial-scale=1.0">
        \\<meta name=author content=Nulo>
        \\<meta property=og:title content="{0s}">
        \\<meta property=og:type content=website>
        \\<meta property=og:url content="https://nulo.in/{1s}.html">
        \\<meta property=og:image content=cowboy.svg>
        \\<link rel=stylesheet href=drip.css>
        \\<link rel=icon href=cowboy.svg>
        \\<title>{0s}</title>
        \\
    , .{ title, src_name });
    // if test "$mirror" = true; then
    // 	echo "<p style=color:darkred>Ojo: este sitio es un espejo (mirror). <a href=https://nulo.in>nulo.in</a> es la fuente.</p>"
    // fi
    if (options.ir_al_inicio) {
        try writer.writeAll("<a href=.>☚ Volver al inicio</a>\n");
    }
    if (options.header) {
        try writer.print(
            \\<header>
            \\<h1>{s}</h1>
            \\<a href="https://gitea.nulo.in/Nulo/sitio/commits/branch/ANTIFASCISTA/{s}">Historial</a>
            \\
        , .{ title, src_name });
        if (options.has_connections) {
            try writer.print(
                \\/
                \\<a href="#conexiones">Conexiones</a>
            , .{});
        }
        try writer.print(
            \\</header>
            \\
        , .{});
    }
}

const Connection = struct {
    linker: []const u8,
    linked: []const u8,
    pub fn make(allocator: std.mem.Allocator, linker: []const u8, linked: []const u8) !Connection {
        return Connection{
            .linker = try allocator.dupe(u8, linker),
            .linked = try allocator.dupe(u8, linked),
        };
    }
};

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    var connections = std.ArrayList(Connection).init(allocator);
    defer connections.deinit();
    defer for (connections.items) |connection| {
        allocator.free(connection.linker);
        allocator.free(connection.linked);
    };

    var page_list = std.ArrayList([]const u8).init(allocator);
    defer page_list.deinit();
    defer for (page_list.items) |item| allocator.free(item);

    var cwd = try std.fs.cwd().openDir(".", .{ .iterate = true });
    defer cwd.close();
    var cwd_iterator = cwd.iterate();

    var build_dir = try cwd.makeOpenPath("build", .{});
    defer build_dir.close();

    while (try cwd_iterator.next()) |entry| {
        if (entry.kind != .File) continue;
        if (endsWith(u8, entry.name, ".md"))
            try scanForConnections(allocator, cwd, &connections, entry.name);
    }

    cwd_iterator = cwd.iterate();

    while (try cwd_iterator.next()) |entry| {
        if (entry.kind != .File) continue;

        // Autocopiarnos :)
        if (endsWith(u8, entry.name, ".zig") or
            endsWith(u8, entry.name, ".md") or
            endsWith(u8, entry.name, ".css") or
            endsWith(u8, entry.name, ".png") or
            endsWith(u8, entry.name, ".jpg") or
            endsWith(u8, entry.name, ".mp4") or
            endsWith(u8, entry.name, ".svg") or
            endsWith(u8, entry.name, ".html"))
        {
            try cwd.copyFile(entry.name, build_dir, entry.name, .{});
        }

        if (endsWith(u8, entry.name, ".md") or endsWith(u8, entry.name, ".gen")) {
            try page_list.append(try allocator.dupe(u8, try stripExtension(entry.name)));
        }

        if (endsWith(u8, entry.name, ".md"))
            try generateMarkdown(allocator, cwd, entry.name, build_dir, connections);
        if (endsWith(u8, entry.name, ".gen"))
            try generateExecutable(allocator, cwd, entry.name, build_dir);
    }

    try generatePageList(allocator, build_dir, page_list.items);
}

const Writer = std.io.BufferedWriter(4096, std.fs.File.Writer).Writer;

fn scanForConnections(
    allocator: std.mem.Allocator,
    cwd: std.fs.Dir,
    connections: *std.ArrayList(Connection),
    file_name: []const u8,
) !void {
    var file = try cwd.openFile(file_name, .{});
    defer file.close();

    const name = try stripExtension(file_name);

    const markdown = try file.readToEndAllocOptions(
        allocator,
        69696969,
        null,
        @alignOf(u32),
        0,
    );
    defer allocator.free(markdown);
    var iter = std.mem.split(u8, markdown, "\n");

    while (iter.next()) |line| {
        var index: usize = 0;
        while (index < line.len) : (index += 1) {
            const rest = line[index..];

            if (std.mem.startsWith(u8, rest, "[[")) {
                if (std.mem.indexOf(u8, rest, "]]")) |end| {
                    const linked = rest[2..end];
                    index += 2 + linked.len + 2 - 1;
                    try connections.append(try Connection.make(allocator, name, linked));
                    continue;
                }
            }
        }
    }
}

fn hackilyTransformHtml(input: []const u8, writer: Writer) !void {
    var iter = std.mem.split(u8, input, "\n");
    while (iter.next()) |line| {
        var index: usize = 0;

        while (index < line.len) : (index += 1) {
            const rest = line[index..];

            const link_marker = "<a h";
            if (std.ascii.startsWithIgnoreCase(rest, link_marker)) {
                index += link_marker.len - 1;
                try writer.writeAll("<a rel='noopener noreferrer' h");
                continue;
            }

            if (std.mem.startsWith(u8, rest, "[[")) {
                if (std.mem.indexOf(u8, rest, "]]")) |end| {
                    const name = rest[2..end];
                    index += 2 + name.len + 2 - 1;
                    try writer.print(
                        \\<a href="{0s}.html">{0s}</a>
                    , .{name});
                    continue;
                }
            }

            try writer.writeByte(line[index]);
        }

        try writer.writeByte('\n');
    }
}

fn generateMarkdown(
    allocator: std.mem.Allocator,
    cwd: std.fs.Dir,
    src_name: []const u8,
    build_dir: std.fs.Dir,
    connections: std.ArrayList(Connection),
) !void {
    var file = try cwd.openFile(src_name, .{});
    defer file.close();
    const markdown = try file.readToEndAllocOptions(
        allocator,
        69696969,
        null,
        @alignOf(u32),
        0,
    );
    defer allocator.free(markdown);
    const html = c.cmark_markdown_to_html(
        markdown.ptr,
        markdown.len,
        c.CMARK_OPT_UNSAFE | c.CMARK_OPT_SMART,
    );
    defer std.c.free(html);

    const output_file_name = try std.fmt.allocPrint(
        allocator,
        "{s}.html",
        .{try stripExtension(src_name)},
    );
    defer allocator.free(output_file_name);

    const is_index = std.ascii.eqlIgnoreCase(src_name, "index.md");

    const title = if (is_index)
        "nulo.in"
    else
        try stripExtension(src_name);

    var connection_count: u16 = 0;
    for (connections.items) |connection| {
        if (std.mem.eql(u8, try stripExtension(src_name), connection.linked))
            connection_count += 1;
    }

    var output = try build_dir.createFile(output_file_name, .{});
    defer output.close();
    try header(output.writer(), title, src_name, .{
        .ir_al_inicio = !is_index,
        .header = !is_index,
        .has_connections = connection_count > 0,
    });
    var buffered_writer = std.io.bufferedWriter(output.writer());
    try hackilyTransformHtml(
        std.mem.span(html),
        buffered_writer.writer(),
    );

    if (connection_count > 0) {
        try buffered_writer.writer().print(
            \\<section id=conexiones>
            \\<h2>⥆ Conexiones ({})</h2>
            \\<ul>
        , .{connection_count});
        for (connections.items) |connection|
            if (std.mem.eql(u8, try stripExtension(src_name), connection.linked))
                try buffered_writer.writer().print(
                    \\<li><a href="{0s}.html">{0s}</a></li>
                , .{connection.linker});
        try buffered_writer.writer().print(
            \\</ul>
            \\</section>
        , .{});
    }
    try buffered_writer.flush();
}

fn generateExecutable(
    allocator: std.mem.Allocator,
    cwd: std.fs.Dir,
    src_name: []const u8,
    build_dir: std.fs.Dir,
) !void {
    _ = cwd;
    const output_file_name = try std.fmt.allocPrint(
        allocator,
        "{s}.html",
        .{try stripExtension(src_name)},
    );
    defer allocator.free(output_file_name);

    const title = try stripExtension(src_name);

    var output = try build_dir.createFile(output_file_name, .{});
    defer output.close();
    try header(output.writer(), title, src_name, .{});

    const executable_name = try std.fmt.allocPrint(allocator, "./{s}", .{src_name});
    defer allocator.free(executable_name);

    // const process = try std.ChildProcess.init(&.{executable_name}, allocator);
    // defer process.deinit();
    // process.stdout_behavior = .Ignore;
    // process.stdout = output;
    // const term = try process.spawnAndWait();
    const result = try std.ChildProcess.exec(.{
        .allocator = allocator,
        .argv = &.{executable_name},
    });
    defer allocator.free(result.stdout);
    defer allocator.free(result.stderr);
    if (result.stderr.len > 0) {
        std.log.err("{s} stderr: {s}", .{ src_name, result.stderr });
    }
    switch (result.term) {
        .Exited => |status| {
            if (status != 0) return error.ProcessFailed;
        },
        else => unreachable,
    }
    try output.writeAll(result.stdout);
}

fn generatePageList(
    allocator: std.mem.Allocator,
    build_dir: std.fs.Dir,
    page_list: []const []const u8,
) !void {
    _ = allocator;
    const name = "Lista de páginas";
    var output = try build_dir.createFile(name ++ ".html", .{});
    defer output.close();
    try header(output.writer(), name, "compilar.zig", .{});

    try output.writeAll("<ul>");
    for (page_list) |page| {
        try output.writer().print(
            \\<li><a href="{0s}.html">{0s}</a></li>
        , .{page});
    }
    try output.writeAll("</ul>");
}

File: compilar.ts
import { copyFile, mkdir, opendir, readFile, readdir, writeFile } from "fs/promises";
import { basename, extname, join } from "path";
import * as commonmark from "commonmark";
import {
  a,
  doctype,
  h1,
  header,
  section,
  li,
  link,
  meta,
  metaUtf8,
  render,
  Renderable,
  title,
  ul,
  h2,
  raw,
  p,
  VirtualElement,
  time,
  article,
  main,
  img,
  script,
  basicElement,
  nav,
  source,
} from "@nulo/html.js";

const div = basicElement("div");

const reader = new commonmark.Parser({ smart: true });
const writer = new commonmark.HtmlRenderer({ safe: false, smart: true });

const dateFormatter = new Intl.DateTimeFormat("es-AR", {
  weekday: "long",
  day: "numeric",
  month: "long",
  year: "numeric",
});

const wikilinkExp = /\[\[(.+?)\]\]/giu;

interface Config {
  sourcePath: string;
  buildPath: string;
}
const config: Config = {
  sourcePath: ".",
  buildPath: "build",
};

const buscadorHtml = await readFile("buscador.htm", "utf-8");

const connections = await scanForConnections(config.sourcePath);

await mkdir(config.buildPath, { recursive: true });

const dir = await readdir(config.sourcePath, { withFileTypes: true });
let pageList: { src: string }[] = [];
for (const entry of dir) {
  if (!entry.isFile()) continue;
  const { name } = entry;
  const extension = extname(name);

  if ([".ts", ".md", ".css", ".js", ".png", ".jpg", ".mp4", ".svg", ".html"].includes(extension)) {
    await copyFile(join(config.sourcePath, name), join(config.buildPath, name));
  }

  if ([".md"].includes(extension) || name.endsWith(".gen.js")) {
    pageList.push({ src: name });
  }
}
await Promise.all(pageList.map(({ src }) => compilePage(config, src)));

await compilePageList(config, pageList);

async function compileFile(
  config: Config,
  sourceFileName: string
): Promise<{ contentHtml: string; image?: Image }> {
  if (extname(sourceFileName) === ".md") {
    const { html: contentHtml, image } = await compileMarkdownHtml(config, sourceFileName);
    return { contentHtml, image };
  } else if (sourceFileName.endsWith(".gen.js")) {
    const contentHtml = await compileJavascript(config, sourceFileName);
    return { contentHtml };
  } else if (sourceFileName.endsWith(".htm"))
    return { contentHtml: await readFile(sourceFileName, "utf-8") };
  else throw false;
}

async function compilePage(config: Config, sourceFileName: string) {
  const name = basename(sourceFileName, extname(sourceFileName));
  const isIndex = name === "index";
  const title = isIndex ? "nulo.ar" : formatNameToPlainText(name);
  const fileConnections = connections.filter(({ linked }) => linked === name);

  const { contentHtml, image } = await compileFile(config, sourceFileName);

  const html = renderHtml(
    ...generateHead(title, name),
    article(
      { itemscope: "", itemtype: "https://schema.org/Article" },
      ...(isIndex ? [] : generateHeader(name, sourceFileName, fileConnections.length > 0, image)),
      main({ itemprop: "articleBody", "data-pagefind-body": "" }, raw(contentHtml)),
      ...generateConnectionsSection(fileConnections)
    )
  );

  const outputPath = join(config.buildPath, name + ".html");
  await writeFile(outputPath, html);
}

// ==============================================
// Get HTML
// ==============================================

type Image = {
  src: string;
  alt: string;
};

async function compileMarkdownHtml(
  config: Config,
  sourceFileName: string
): Promise<{ html: string; image?: Image }> {
  let markdown = await readFile(join(config.sourcePath, sourceFileName), "utf-8");

  let image;
  if (markdown.startsWith("!!")) {
    const node = reader.parse(markdown.slice(1, markdown.indexOf("\n")));
    const imageNode = node.firstChild?.firstChild;
    if (!imageNode || !imageNode.destination)
      throw new Error("Intenté parsear un ^!! pero no era una imágen");
    if (!imageNode.firstChild?.literal) console.warn(`El ^!! de ${sourceFileName} no tiene alt`);

    image = {
      src: imageNode.destination,
      alt: imageNode.firstChild?.literal || "",
    };
    markdown = markdown.slice(markdown.indexOf("\n"));
  }

  const parsed = reader.parse(markdown);
  const markdownHtml = writer.render(parsed);

  checkLinks(sourceFileName, markdownHtml);
  const contentHtml = await hackilyTransformHtml(markdownHtml);
  return { html: contentHtml, image };
}

async function compileJavascript(config: Config, sourceFileName: string): Promise<string> {
  const fn = await import("./" + join(config.sourcePath, sourceFileName));
  return await fn.default();
}

// ==============================================
// Generated HTML
// ==============================================

function renderHtml(...world: Renderable[]): string {
  return `<!doctype html><html lang="es">` + render(...world) + "</html>";
}

function generateHead(titlee: string, outputName: string): Renderable[] {
  // TODO: deshardcodear og:url
  return [
    metaUtf8,
    meta({
      name: "viewport",
      content: "width=device-width, initial-scale=1.0",
    }),
    meta({ name: "author", content: "Nulo" }),
    meta({ property: "og:title", content: titlee }),
    meta({ property: "og:type", content: "website" }),
    meta({ property: "og:url", content: `https://nulo.ar/${outputName}.html` }),
    meta({ property: "og:image", content: "cowboy.svg" }),
    link({ rel: "stylesheet", href: "drip.css" }),
    link({ rel: "icon", href: "cowboy.svg" }),
    title(titlee),
  ];
}

function formatDate(dateish: Dateish, upper: boolean = false): string {
  const date = new Date(dateish.year, dateish.month - 1, dateish.day);
  const formatted = dateFormatter.format(date);
  if (upper) {
    // no le digan a la policía del unicode!
    return formatted[0].toUpperCase() + formatted.slice(1);
  } else return formatted;
}

interface Dateish {
  year: number;
  month: number;
  day: number;
}
function dateishToString({ year, month, day }: Dateish): string {
  return `${year}-${String(month).padStart(2, "0")}-${String(day).padStart(2, "0")}`;
}

type TitleMetadata =
  | {
      // title puede tener length == 0 y por lo tanto ser falseish
      title: string;
      date?: Dateish;
    }
  | { date: Dateish };
function parseName(name: string): TitleMetadata {
  const titleWithDate = /^((?<year>\d{4})-(?<month>\d{2})-(?<day>\d{2}))? ?(?<title>.*)$/;

  const found = name.match(titleWithDate);
  if (!found || !found.groups) throw new Error("Algo raro pasó");
  const { title } = found.groups;

  const date =
    (found.groups.year && {
      year: parseInt(found.groups.year),
      month: parseInt(found.groups.month),
      day: parseInt(found.groups.day),
    }) ||
    undefined;
  // no definir title si es length == 0
  if (!title && date) return { date };
  return { title, date };
}

function dateishToElement(
  dateish: Dateish,
  { itemprop, upper }: { itemprop?: string; upper?: boolean } = {}
): VirtualElement {
  return time(
    { datetime: dateishToString(dateish), ...(itemprop ? { itemprop } : {}) },
    formatDate(dateish, upper)
  );
}

function formatNameToInline(name: string): Renderable[] {
  const parsed = parseName(name);
  if ("title" in parsed) {
    const { title, date } = parsed;
    return [title, ...(date ? [` (`, dateishToElement(date), `)`] : [])];
  } else {
    return [dateishToElement(parsed.date, { upper: true })];
  }
}
function formatNameToPlainText(name: string): string {
  const parsed = parseName(name);
  if ("title" in parsed) {
    const { title, date } = parsed;
    return title + (date ? ` (${formatDate(date)})` : "");
  } else {
    return formatDate(parsed.date, true);
  }
}

function generateHeader(
  name: string,
  sourceCodePath: string,
  linkConexiones = false,
  image?: Image
): Renderable[] {
  const parsedTitle = parseName(name);
  return [
    nav(a({ href: "." }, "☚ Volver al inicio"), raw(buscadorHtml)),
    header(
      ...(image ? [img({ ...image, itemprop: "image" })] : []),
      ...("title" in parsedTitle
        ? [
            h1(parsedTitle.title),
            ...(parsedTitle.date
              ? [
                  dateishToElement(parsedTitle.date, {
                    itemprop: "datePublished",
                  }),
                  " / ",
                ]
              : []),
          ]
        : [
            h1(
              dateishToElement(parsedTitle.date, {
                itemprop: "datePublished",
                upper: true,
              })
            ),
          ]),
      a(
        {
          href: `https://gitea.nulo.in/Nulo/sitio/commits/branch/ANTIFASCISTA/${sourceCodePath}`,
        },
        "Historial"
      ),
      ...(linkConexiones ? [" / ", a({ href: "#conexiones" }, "Conexiones")] : [])
    ),
  ];
}

function generateConnectionsSection(fileConnections: Connection[]): Renderable[] {
  return fileConnections.length > 0
    ? [
        section(
          { id: "conexiones" },
          h2(`⥆ Conexiones (${fileConnections.length})`),
          ul(...fileConnections.map(({ linker }) => li(internalLink(linker))))
        ),
      ]
    : [];
}

async function compilePageList(config: Config, pageList: { src: string }[]) {
  const name = "Lista de páginas";
  const outputPath = join(config.buildPath, name + ".html");
  const html = renderHtml(
    ...generateHead(name, name),
    ...generateHeader(name, "compilar.ts"),
    ul(
      ...pageList
        .map(({ src: name }) => basename(name, extname(name)))
        .sort((a, b) => a.localeCompare(b, "es", { sensitivity: "base" }))
        .map((name) => li(internalLink(name)))
    )
  );
  await writeFile(outputPath, html);
}

// ==============================================
// Conexiones
// ==============================================

interface Connection {
  linked: string;
  linker: string;
}

async function scanForConnections(sourcePath: string): Promise<Connection[]> {
  const dir = await opendir(sourcePath);
  let connections: Connection[] = [];
  for await (const entry of dir) {
    const extension = extname(entry.name);
    if (extension === ".md") {
      const name = basename(entry.name, ".md");
      const file = await readFile(join(config.sourcePath, entry.name), "utf-8");
      for (const [, linked] of file.matchAll(wikilinkExp)) {
        connections.push({ linked, linker: name });
      }
    }
  }
  return connections;
}

// ==============================================
// Markdown utils
// ==============================================

async function hackilyTransformHtml(html: string): Promise<string> {
  html = html
    .replaceAll("<a h", '<a rel="noopener noreferrer" h')
    .replaceAll(wikilinkExp, (_, l) => render(internalLink(l)));
  for (const [match, archivo] of html.matchAll(/<nulo-sitio-reemplazar-con archivo="(.+?)" \/>/g)) {
    html = html.replace(match, (await compileFile(config, archivo)).contentHtml);
  }
  return html;
}

function checkLinks(srcName: string, html: string) {
  const matches = html.matchAll(wikilinkExp);
  const list = pageList.map(({ src }) => basename(src, extname(src)));
  if (!matches) return;
  for (const match of matches) {
    if (!list.some((n) => n === match[1])) {
      console.warn(`${srcName} linkea a ${match[1]}, pero no existe`);
    }
  }
}

// ==============================================
// Linking
// ==============================================

function internalLink(path: string): VirtualElement {
  const href = encodeURI(`./${path}.html`);
  return a({ href }, ...formatNameToInline(path));
}
